src/lin_diffEq_HHL.jl

```diff
@@ -19,106 +19,91 @@ export QuEuler, QuLeapfrog, QuAB2, QuAB3, QuAB4
+    A
 end
 """
     Explicit Linear Multistep Methods
 """
-QuEuler = alg_lin_diffeq(1,[1],[1]);
-QuLeapfrog = alg_lin_diffeq(2,[0 1],[2 0])
-QuAB2 = alg_lin_diffeq(2,[1 0], [1.5 -0.5])
-QuAB3 = alg_lin_diffeq(3,[1 0 0], [23/12 -16/12 5/12])
-QuAB4 = alg_lin_diffeq(4,[1 0 0 0], [55/24 -59/24 37/24 -9/24])
-
-
-function prepare_init_state(tspan::Tuple,x::Vector,h::Float64,g::Function,alg::alg_lin_diffeq)
-  init_state = x;
-  N_t = Int(round(2*(tspan[2] - tspan[1])/h + 3))
-  b = similar(g(1))
-  for i = 1:N_t
-   if i < (N_t+1)/2 -1
-      b = bval(alg,h*(i-1) + tspan[1],h,g)
-      init_state = [init_state;h*b]
-    else
-      init_state = [init_state;zero(b)]
-   end
-
-  end
-  init_state = [init_state;zero(init_state)]
-  init_state
+QuEuler = AlgQuDiff(1,[1],[1]);
+QuLeapfrog = AlgQuDiff(2,[0 1],[2 0])
+QuAB2 = AlgQuDiff(2,[1 0], [1.5 -0.5])
+QuAB3 = AlgQuDiff(3,[1 0 0], [23/12 -16/12 5/12])
+QuAB4 = AlgQuDiff(4,[1 0 0 0], [55/24 -59/24 37/24 -9/24])
+
+
+function prepare_init_state(tspan::Tuple,x::Vector,h::Float64,g::Function,alg::AlgQuDiff)
+    N_t = Int(round(2*(tspan[2] - tspan[1])/h + 3))
+    sz, = size(g(1))
+    init_state = zeros(ComplexF64,2*(N_t + 1)*sz)
+    #inital value
+    init_state[1:sz] = x
+    for i in 2:(N_t+1)/2 - 1
+        b = bval(alg,h*(i - 1) + tspan[1],h,g)
+        init_state[Int(sz*(i - 1) + 1):Int(sz*(i))] = h*b
+    end
+    init_state
 end
 
-function Array_QuDiff(tspan::Tuple,h::Float64,g::Function,alg::alg_lin_diffeq)
-  N_t = Int(round(2*(tspan[2] - tspan[1])/h + 3))
-  sz = size(g(1))
-  I_mat = Matrix{Float64}(I, size(g(1)));
-  A_ = I_mat;
-  # Generates First two rows
-  A_ = [A_ zeros(sz[1],sz[1]*(N_t));-1*(I_mat + h*g(tspan[1])) I_mat zeros(sz[1],sz[1]*(N_t-1))]
-  #Generates additional rows based on k - step
-  if alg.step > 1
-    A_ = [A_; Aval(QuAB2,2*h + tspan[1],h,g) zeros(sz[1],sz[1]*(N_t-2))]
-  end
-
-  for i = 3:alg.step-1
-    tA_ = Aval(QuAB2,(i-1)*h + tspan[1],h,g)
-    tA_ = [zeros(sz[1],sz[1]*(i-2)) tA_]
-    A_ = [A_; tA_ zeros(sz[1],sz[1]*(N_t-i))]
-  end
-
-  if alg.step > 2
-    A_ = [A_;Aval(alg,(alg.step)*h + tspan[1],h,g) zeros(sz[1],sz[1]*(N_t-alg.step)) ]
-  end
-
-  for i = alg.step+1: (N_t + 1)/2 -2
-    tA_ = Aval(alg,(i-1)*h + tspan[1],h,g)
-    A_= [A_;zeros(sz[1],Int(sz[1]*(i-alg.step))) tA_ zeros(sz[1],Int(sz[1]*(N_t-i)))]
-   end
-  #Generates half mirroring matrix
-  for i = (N_t + 1)/2  : N_t
-    A_ = [A_; zeros(sz[1],Int(sz[1]*(i-2))) -1*I_mat I_mat zeros(sz[1],Int(sz[1]*(N_t + 1 -i)))]
-  end
-  A_ = [A_; zeros(sz[1],sz[1]*(N_t-1)) -1*I_mat I_mat];
-
-  A_ = [zero(A_) A_;A_' zero(A_)]
-  A_
+function array_qudiff(tspan::Tuple,h::Float64,g::Function,alg::AlgQuDiff)
+    sz, = size(g(1))
+    i_mat = Matrix{Float64}(I, size(g(1)))
+    N_t = Int(round(2*(tspan[2] - tspan[1])/h + 3))
+    A_ = zeros(ComplexF64,(N_t + 1)*sz,(N_t + 1)*sz)
+    # Generates First two rows
+    A_[1:sz, 1:sz] = i_mat
+    A_[sz + 1:2*sz,1:sz*2] = [-1*(i_mat + h*g(tspan[1])) i_mat]
```

Comment:
I think you could assign this matrix separately to avoid extra allocation, concat will allocate new matrix instead of doing the assignment in-place.

Reply:
I am not quite sure about how to do this.

Reply:
this is a MWE:


What you are doing here:

```julia
julia> M = rand(2, 2)
2×2 Array{Float64,2}:
 0.995624  0.888867
 0.282985  0.294511

julia> [M M]
2×4 Array{Float64,2}:
 0.995624  0.888867  0.995624  0.888867
 0.282985  0.294511  0.282985  0.294511

julia> A = rand(4, 4)
4×4 Array{Float64,2}:
 0.73779   0.441079  0.637734   0.133061
 0.654057  0.600815  0.686549   0.300993
 0.864947  0.517726  0.0534451  0.766352
 0.199765  0.719381  0.39159    0.022419

julia> A[1:2, 1:4] = [M M]
2×4 Array{Float64,2}:
 0.995624  0.888867  0.995624  0.888867
 0.282985  0.294511  0.282985  0.294511
```

But in fact you can directly 

```julia
A[1:2, 1:2] = M
A[1:2, 3:4] = M
```

the concat `[M M]` will create a new matrix, instead of doing the assignment in-place. And don't forget to add `@inbounds` btw.

Reply:
Oh, I see. Thanks.